extra_geom/base.py

```diff
@@ -791,3 +798,135 @@ def offset(self, shift, *, modules=np.s_[:], tiles=np.s_[:]):
+            radians.
+        """
+        rot = np.asarray(angles)
+        if rot.shape[-1] != 3:
+            raise ValueError('Rotation angles must be 3d (x, y, z). '
+                             f'Last dimension was {rot.shape[-1]}')
+        if center is not None and np.asarray(center).shape != rot.shape:
+            raise ValueError('Rotation center must have same shape as angles')
+
+        ntiles = len(max((m for m in self.modules), key=len))
+        all_rot = np.zeros((len(self.modules), ntiles, 3), dtype=rot.dtype)
+        sel_rot = all_rot[modules, tiles]
+
+        refs = np.zeros((len(self.modules), ntiles, 3), dtype=np.float64)
+        sel_refs = refs[modules, tiles]
+
+        if center is None:
+            if rot.shape[:-1] == sel_rot.shape[:2] or tiles != np.s_[:]:
+                # Per-tile rotation
+                refs[modules, tiles] = np.array(
+                    [[t.centre() for t in m[tiles]]
+                    for m in self.modules[modules]]
+                )
+            elif rot.shape[:-1] == sel_rot.shape[:1] or modules != np.s_[:]:
+                # Per-module rotation
+                mods = []
+                for module in self.modules[modules]:
+                    center = np.mean([t.centre() for t in module], axis=0)
+                    mods.append(np.array([center for _ in module[tiles]]))
+                refs[modules, tiles] = np.array(mods)
+        else:
+            center = np.asarray(center)
+            if center.shape[:-1] == sel_refs.shape[:2]:
+                # Per-tile offsets
+                sel_refs[:] = center
+            elif center.shape[:-1] == sel_refs.shape[:1]:
+                # Per-module offsets - broadcast across tiles
+                sel_refs[:] = center[:, np.newaxis]
+            elif center.shape[:-1] == ():
+                # Single shift - broadcast across modules and tiles
+                sel_refs[:] = center
+            else:
+                raise ValueError(
+                    f"Got {center.shape[:-1]} coordinates. Expected either a "
+                    f"single coordinate (), a coordinate per module "
+                    f"{sel_refs.shape[:1]} or a coordinate per tile "
+                    f"{sel_refs.shape[:2]}"
+                )
+
+        if rot.shape[:-1] == sel_rot.shape[:2]:
+            # per-tile rotations
+            sel_rot[:] = rot
+        elif rot.shape[:-1] == sel_rot.shape[:1]:
+            # per-module rotations - broadcast across tiles
+            sel_rot[:] = rot[:, np.newaxis]
+        elif rot.shape[:-1] == ():
+            # single rotation - broadcast across modules and tiles
+            sel_rot[:] = rot
+        else:
+            raise ValueError(
+                f"Got {rot.shape[:-1]} coordinates. Expected either a single "
+                f"coordinate (), a coordinate per module {sel_rot.shape[:1]} "
+                f"or a coordinate per tile {sel_rot.shape[:2]}"
```

Comment:
I wouldn't call these coordinates - is there a term for a set of 3 angles for 3D rotation? Or do people with more of a maths background call them coordinates?

Reply:
In analytical mechanics people may refer to angles as "generalized coordinates". For example, Euler angles or polar angles could be referred to as coordinates in this context (https://en.wikipedia.org/wiki/Generalized_coordinates). I would not see this as a problem. Even if it is, I think it is clear what is meant by context, right?

Reply:
If you're passing both angles and centre coordinates, and you get an error talking about 'coordinates', I'd probably guess it meant the centre coordinates rather than the angles. It's probably not that hard to work out, I was just wondering if we could find a clearer wording.

Reply:
I changed `coordinates` to `rotation sequences`. Is it clearer? Or do you have another suggestion?

Reply:
Thanks, I think that's clear enough.